### PR TITLE
Hide output if it is much longer than expected to avoid browser crashes

### LINKED ISSUE #4265: Too much feedback is not always properly stopped from showing up
In most cases, Dodona automatically stops printing the output of a specific submission if this becomes too much. However, this seems to be judge-specific behaviour and is not handled properly in all cases causing some browsers to crash.

See this submission link for an example of the issue: https://dodona.ugent.be/nl/submissions/12839569/

### PR DESCRIPTION
**Pull request by jorg-vr**
This pull request hides very long submission output to avoid browser crashes. This solves #4265 

To much output is defined as 5000  characters more than the expected output. (But this is open for discussion)

image

Closes #4265.

### CODE REVIEW

File: app/helpers/renderers/text_differ.rb [hunk 1]
```diff
@@ -21,6 +21,11 @@ def initialize(generated, expected)
                 end
               end
             end
+    return unless @generated_linecount > @expected_linecount + 100
+
+    @to_many_lines = @generated_linecount
```
Comment (chvp) [inline app/helpers/renderers/text_differ.rb:26]:
```suggestion
    @too_many_lines = @generated_linecount
```

File: app/helpers/renderers/text_differ.rb [hunk 2]
```diff
@@ -47,6 +52,8 @@ def unified
         nil
       end
     end.html_safe
+    cut_off_message(builder) if @to_many_lines
```
Comment (chvp) [inline app/helpers/renderers/text_differ.rb:55]:
```suggestion
    cut_off_message(builder) if @too_many_lines
```

File: app/helpers/renderers/text_differ.rb [hunk 3]
```diff
@@ -237,4 +244,10 @@ def diff_strings(generated, expected)
     exp_result += '</strong>' if in_exp_strong
     [gen_result, exp_result]
   end
+
+  def cut_off_message(builder)
+    builder.div(class: 'callout callout-warning') do
+      builder << I18n.t('submissions.show.cut_off_message', generated: @to_many_lines, count: @expected_linecount)
```
Comment (chvp) [inline app/helpers/renderers/text_differ.rb:250]:
```suggestion
      builder << I18n.t('submissions.show.cut_off_message', generated: @too_many_lines, count: @expected_linecount)
```

File: app/helpers/renderers/text_differ.rb [hunk 4]
```diff
@@ -21,6 +21,11 @@ def initialize(generated, expected)
                 end
               end
             end
+    return unless @generated_linecount > @expected_linecount + 100
```
Comment (chvp) [inline app/helpers/renderers/text_differ.rb:24]: A return in the constructor feels weird to me.

**Review by chvp — CHANGES_REQUESTED** [review]
I think we should tackle lines that are far too long as well. In the case of #4265, we could imagine a student not adding a newline character. This would then result in one huge line, and the same problem would occur. Cutting off the output if there are 2000 (or something like that) characters more than expected would solve both cases I think.

The interaction with the simplified_table functionality also feels weird to me somehow. In the case where there is such a difference between generated and expected, I would also expect to also use the simplified table, but that's not very clear from the code.

File: app/helpers/renderers/text_differ.rb (hunk 4 above)
Comment (jorg-vr) [reply app/helpers/renderers/text_differ.rb:24]: The constructor is also just a function...

This was also forced by our linter (Which prefers guard clauses over a regular if)

**Review by chvp — APPROVED** [review]

**Review by bmesuere — APPROVED** [review]

**Review by niknetniko — APPROVED** [review]

**Pull request merged**